s3transfer/download.py

```diff
@@ -0,0 +1,311 @@
+        temp_f = self._get_io_file_handle(temp_filename, transfer_coordinator)
+
+        # Determine the number of parts
+        part_size = self._config.multipart_chunksize
+        num_parts = int(
+            math.ceil(transfer_future.meta.size / float(part_size)))
+
+        # Before starting any tasks for downloads set up a cleanup function
+        # that will make sure that the temporary file always gets
+        # cleaned up.
+        transfer_coordinator.add_failure_cleanup(
+            self._osutil.remove_file, temp_filename)
+
+        ranged_downloads = []
+        for i in range(num_parts):
+            # Calculate the range parameter
+            range_parameter = self._calculate_range_param(
+                part_size, i, num_parts)
+
+            # Inject the Range parameter to the parameters to be passed in
+            # as extra args
+            extra_args = {'Range': range_parameter}
+            extra_args.update(call_args.extra_args)
+            # Submit the ranged downloads
+            ranged_downloads.append(
+                self._executor.submit(
+                    GetObjectTask(
+                        transfer_coordinator=transfer_coordinator,
+                        main_kwargs={
+                            'client': self._client,
+                            'bucket': call_args.bucket,
+                            'key': call_args.key,
+                            'f': temp_f,
+                            'extra_args': extra_args,
+                            'callbacks': progress_callbacks,
+                            'max_attempts': self._config.num_download_attempts,
+                            'io_executor': self._io_executor,
+                            'start_index': i * part_size
+                        }
+                    )
+                )
+            )
+        # Send the necessary tasks to complete the download.
+        self._complete_download(
+            temp_f, call_args.fileobj, transfer_coordinator,
+            ranged_downloads, done_callbacks
+        )
+
+    def _get_io_file_handle(self, filename, transfer_coordinator):
+        f = self._osutil.open(filename, 'wb')
+        transfer_coordinator.add_failure_cleanup(f.close)
+        return f
+
+    def _complete_download(self, f, final_filename,
+                           transfer_coordinator, download_futures,
+                           done_callbacks):
+        # A task to rename the file from the temporary file to its final
+        # location. This should be the last task needed to complete the
+        # download.
+        rename_task = IORenameFileTask(
+            transfer_coordinator=transfer_coordinator,
+            main_kwargs={
+                'f': f,
+                'final_filename': final_filename,
+                'osutil': self._osutil
+            },
+            done_callbacks=done_callbacks,
+            is_final=True
+        )
+        submit_rename_task = functools.partial(
```

Comment:
I'd prefer not to use partials in situations like this.  It usually makes the logs/tracebacks harder to follow.  In this case, the concept of submitting a task after another task finishes seems like it's worth extracting into it's own class.